packages/fx-core/src/component/driver/file/createOrUpdateDebugProfile.ts

```diff
@@ -0,0 +1,139 @@
+import * as utils from "util";
+import { internalOutputNames } from "../teamsApp/create";
+import { isCommentObject } from "../../../core/middleware/utils/debug/debugV3MigrationUtils";
+
+const actionName = "file/createOrUpdateDebugProfile";
+const helpLink = "https://aka.ms/teamsfx-actions/file-createOrUpdateDebugProfile";
+const launchSettingsFilePath = "./Properties/launchSettings.json";
+const launchUrlTemplate =
+  "https://teams.microsoft.com/l/app/%s?installAppPackage=true&webjoin=true&appTenantId=%s";
+
+@Service(actionName) // DO NOT MODIFY the service name
+export class CreateOrUpdateDebugProfileDriver implements StepDriver {
+  description = getLocalizedString("driver.file.createOrUpdateDebugProfile.description");
+
+  @hooks([
+    addStartAndEndTelemetry(actionName, actionName),
+    updateProgress(getLocalizedString("driver.file.progressBar.createOrUpdateDebugProfile")),
+  ])
+  public async run(
+    args: CreateOrUpdateDebugProfileArgs,
+    context: DriverContext
+  ): Promise<Result<Map<string, string>, FxError>> {
+    return wrapRun(async () => {
+      const result = await this.handler(args, context);
+      return result.output;
+    });
+  }
+
+  @hooks([
+    addStartAndEndTelemetry(actionName, actionName),
+    updateProgress(getLocalizedString("driver.file.progressBar.createOrUpdateDebugProfile")),
+  ])
+  public async execute(
+    args: CreateOrUpdateDebugProfileArgs,
+    ctx: DriverContext
+  ): Promise<ExecutionResult> {
+    let summaries: string[] = [];
+    const outputResult = await wrapRun(async () => {
+      const result = await this.handler(args, ctx);
+      summaries = result.summaries;
+      return result.output;
+    });
+    return {
+      result: outputResult,
+      summaries,
+    };
+  }
+
+  private async handler(
+    args: CreateOrUpdateDebugProfileArgs,
+    context: DriverContext
+  ): Promise<{
+    output: Map<string, string>;
+    summaries: string[];
+  }> {
+    try {
+      const launchSettingsPath = getAbsolutePath(launchSettingsFilePath, context.projectPath);
+      if (!(await fs.pathExists(launchSettingsPath))) {
+        await fs.writeFile(launchSettingsPath, "{}", "utf-8");
+      }
+      const launchSettingsContent = await fs.readFile(launchSettingsPath, "utf-8");
+      const data = parse(launchSettingsContent.toString());
+      if (!isCommentObject(data)) {
+        throw new UserError(
+          "LaunchSettingsFileInvalid",
+          getLocalizedString(
+            "driver.file.createOrUpdateDebugProfile.launchSettingsFileInvalid",
+            launchSettingsPath
+          ),
+          helpLink
+        );
+      }
+      let launchUrl = utils.format(
+        launchUrlTemplate,
+        args.appId,
+        process.env[internalOutputNames.teamsAppTenantId]
+      );
+      if (args.loginHint === undefined || args.loginHint === true) {
+        const tokenObjectRes = await context.m365TokenProvider.getJsonObject({
+          scopes: AppStudioScopes,
```

Comment:
What if user has not signed in? Prefer to return error instead of popup signin window.

Reply:
In vs, user need to choose m365 before running prepare Teams app. And there is no logic to pop up login page in running prepare Teams app. So I think we do not need to worry this.